app/models/play_group.rb

```diff
@@ -0,0 +1,194 @@
+        highest_index = 0
+        game_selection.each_with_index do |grouping,i|
+          if @group_scores[@group_scores.index{|i| i.keys.first == grouping[0]}].values[0][grouping[1]+1][1] > next_highest[2][1] 
+            next_highest = [grouping[0],grouping[1]+1,@group_scores[@group_scores.index{|i| i.keys.first == grouping[0]}].values[0][grouping[1]+1]]
+            highest_index = i 
+          end
+        end         
+        game_selection[highest_index] = next_highest
+      end
+      rand = [true,false].sample
+      pp "#{rand} #{game_selection[0]}"
+      top_combo = game_selection if top_combo.empty? || game_selection.map{|group| group[2][1]}.sum > top_combo.map{|group| group[2][1]}.sum || ((game_selection.map{|group| group[2][1]}.sum == top_combo.map{|group| group[2][1]}.sum) && rand)
+    end
+    top_combo      
+
+    
+  end
+
+  def set_collection(params)
+    collection_ids = params.select{ |k, v| (k.start_with? "collection") && (v == "1") }.keys.map{|k| k.tr("collection","").to_i}
+    Collection.where(user_id:collection_ids)
+  end
+
+  def set_games(params)
+    @games = Game.where(id: @collection.pluck(:game_id)).by_players(params[:group_size_min],params[:group_size_max],params[:rec]).by_weight(params[:min_weight],params[:max_weight]).play_time(params[:min_length],params[:max_length], params[:max_length])
+  end
+
+  def set_group_sizes(params)
+    if (params[:group_size_min] != params[:group_size_max])
+      @group_sizes = *(params[:group_size_min].to_i..params[:group_size_max].to_i)
+    else
+      @group_sizes = [params[:group_size_min].to_i,params[:group_size_max].to_i]
+    end
+  end
+
+  def group_error_check(params)
+    (params[:group_size_min].to_i > params[:group_size_max].to_i) || (params[:group_size_min].to_i > player_count) || !GroupCheck.check_group(@group_sizes,player_count)
+  end
+
+  def set_groups(params)
+    (params[:group_size_min].to_i..params[:group_size_max].to_i).flat_map{|group_size| @players.map{ |player| player.id}.combination(group_size).to_a}
+  end
+
+  def get_group_combinations
+    @group_combinations
+  end
+
+  def combine_groups(params)
+    combos = []
+    for a in (player_count/params[:group_size_max].to_i..player_count/params[:group_size_min].to_i) do
+      new_combos = @groups.repeated_permutation(a).to_a
+      combos = combos.concat(new_combos)
+      pp a
+    end
+    validate_groups(combos)
+  end
+
+  def validate_groups(combos)
+    combos.select{ |a| a.flatten.uniq.length == player_count()}.select{ |b| b.flatten == b.flatten.uniq}.map{ |c| c.sort }.uniq
+  end
+  
+
+  def set_options(params)
+    Game.all.play_time(params[:min_length], params[:max_length], params[:time]).by_weight(params[:min_weight],params[:max_weight]).by_players(params[:group_size_min],params[:group_size_max])
+  end
+
+  def player_count
+    @players.length
+  end
+
+  def set_player_scores
+    @ratings = Rating.where user_id: @players, game_id: @games
```

Comment:
Calling all ratings here once and finding within this collection is a lot quicker than querying the ratings table for each player/game combo